Mathlib/Analysis/Convex/EGauge.lean

```diff
@@ -0,0 +1,195 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Analysis.Seminorm
+
+/-!
+# The Minkowski functional, normed field version
+
+In tihs file we define `(egauge 𝕜 s ·)`
+to be the Minkowski functional (gauge) of the set `s`
+in a topological vector space `E` over a normed field `𝕜`,
+as a function `E → ℝ≥0∞`.
+
+It is defined as the infimum of the norms of `c : 𝕜` such that `x ∈ c • s`.
+In particular, for `𝕜 = ℝ≥0` this definition gives an `ℝ≥0∞`-valued version of `gauge`
+defined in `Mathlib/Analysis/Convex/Gauge.lean`.
+
+Currently, we can't reuse results about `egauge` for `gauge`,
+because we lack a theory of normed semifields.
+-/
+
+open Set Filter Metric
+open scoped Topology Pointwise ENNReal NNReal
+
+section SMul
+
+/-- The Minkowski functional for vector spaces over normed fields.
+Given a set `s` in a vector space over a normed field `𝕜`,
+`egauge s` is the functional which sends `x : E`
+to the infimum of `‖c‖₊` over `c` such that `x` belongs to `s` scaled by `c`.
+
+The definition only requires `𝕜` to have a `NNNorm` instance
+and `(· • ·) : 𝕜 → E → E` to be defined.
+This way the definition applies, e.g., to `𝕜 = ℝ≥0`.
+For `𝕜 = ℝ≥0`, the function is equal (up to conversion to `ℝ`)
+to the usual Minkowski functional defined in `gauge`. -/
+noncomputable def egauge (𝕜 : Type*) [NNNorm 𝕜] {E : Type*} [SMul 𝕜 E] (s : Set E) (x : E) : ℝ≥0∞ :=
+  ⨅ (c : 𝕜) (_ : x ∈ c • s), ‖c‖₊
+
+variable (𝕜 : Type*) [NNNorm 𝕜] {E : Type*} [SMul 𝕜 E] {c : 𝕜} {s t : Set E} {x : E} {r : ℝ≥0∞}
+
+@[mono, gcongr]
+lemma egauge_anti (h : s ⊆ t) (x : E) : egauge 𝕜 t x ≤ egauge 𝕜 s x :=
+  iInf_mono fun _c ↦ iInf_mono' fun hc ↦ ⟨smul_set_mono h hc, le_rfl⟩
+
+@[simp] lemma egauge_empty (x : E) : egauge 𝕜 ∅ x = ∞ := by simp [egauge]
+
+variable {𝕜}
+
+lemma egauge_le_of_mem_smul (h : x ∈ c • s) : egauge 𝕜 s x ≤ ‖c‖₊ := iInf₂_le c h
+
+lemma le_egauge_iff : r ≤ egauge 𝕜 s x ↔ ∀ c : 𝕜, x ∈ c • s → r ≤ ‖c‖₊ := le_iInf₂_iff
+
+lemma egauge_eq_top : egauge 𝕜 s x = ∞ ↔ ∀ c : 𝕜, x ∉ c • s := by simp [egauge]
+
+lemma egauge_lt_iff : egauge 𝕜 s x < r ↔ ∃ c : 𝕜, x ∈ c • s ∧ ‖c‖₊ < r := by
+  simp [egauge, iInf_lt_iff]
+
+end SMul
+
+section SMulZero
+
+variable (𝕜 : Type*) [NNNorm 𝕜] [Nonempty 𝕜] {E : Type*} [Zero E] [SMulZeroClass 𝕜 E]
+  {c : 𝕜} {s t : Set E} {x : E} {r : ℝ≥0∞}
+
+@[simp] lemma egauge_zero_left_eq_top : egauge 𝕜 0 x = ∞ ↔ x ≠ 0 := by
+  simp [egauge_eq_top]
+
+@[simp] alias ⟨_, egauge_zero_left⟩ := egauge_zero_left_eq_top
+
+end SMulZero
+
+section Module
+
+variable {𝕜 : Type*} [NormedDivisionRing 𝕜] {α E : Type*} [AddCommGroup E] [Module 𝕜 E]
+    {c : 𝕜} {s t : Set E} {x y : E} {r : ℝ≥0∞}
+
+lemma egauge_le_of_smul_mem_of_ne (h : c • x ∈ s) (hc : c ≠ 0) : egauge 𝕜 s x ≤ ‖c‖₊⁻¹ := by
+  rw [← nnnorm_inv]
+  exact egauge_le_of_mem_smul <| (mem_inv_smul_set_iff₀ hc _ _).2 h
+
+lemma egauge_le_of_smul_mem (h : c • x ∈ s) : egauge 𝕜 s x ≤ (↑‖c‖₊)⁻¹ := by
```

Comment:
maybe
```suggestion
/-- When the inversion happens in `ℝ≥0∞`, there is no degenerate case. -/
lemma egauge_le_of_smul_mem (h : c • x ∈ s) : egauge 𝕜 s x ≤ (↑‖c‖₊)⁻¹ := by
```
or adding an explicit cast on `egauge_le_of_smul_mem_of_ne` to make the difference more obvious.

Reply:
I added both comments and explicit casts.